hyperlink/_url.py

```diff
@@ -982,6 +1018,60 @@ def from_text(cls, text):
         return cls(scheme, host, path, query, fragment, port,
                    rooted, userinfo, uses_netloc)
 
+    def normalize(self, scheme=True, host=True, path=True, query=True,
+                  fragment=True):
+        """Return a new URL object with several standard normalizations
+        applied:
+
+        * Decode unreserved characters (`RFC 3986 2.3`_)
+        * Uppercase remaining percent-encoded octets (`RFC 3986 2.1`_)
+        * Convert scheme and host casing to lowercase (`RFC 3986 3.2.2`_)
+        * Resolve any "." and ".." references in the path (`RFC 3986 6.2.2.3`_)
+        * Ensure an ending slash on URLs with an empty path (`RFC 3986 6.2.3`_)
+
+        All are applied by default, but normalizations can be disabled
+        per-part by passing `False` for that part's corresponding
+        name.
+
+        Args:
+           scheme (bool): Convert the scheme to lowercase
+           host (bool): Convert the host to lowercase
+           path (bool): Normalize the path (see above for details)
+           query (bool): Normalize the query string
+           fragment (bool): Normalize the fragment
+
+        >>> url = URL.from_text(u'Http://example.COM/a/../b/./c%2f?%61')
+        >>> print(url.normalize().to_text())
+        http://example.com/b/c%2F?a
+
+        .. _RFC 3986 3.2.2: https://tools.ietf.org/html/rfc3986#section-3.2.2
+        .. _RFC 3986 2.3: https://tools.ietf.org/html/rfc3986#section-2.3
+        .. _RFC 3986 2.1: https://tools.ietf.org/html/rfc3986#section-2.1
+        .. _RFC 3986 6.2.2.3: https://tools.ietf.org/html/rfc3986#section-6.2.2.3
+        .. _RFC 3986 6.2.3: https://tools.ietf.org/html/rfc3986#section-6.2.3
+
+        """
+        # TODO: userinfo?
```

Comment:
This should reference a URL to a ticket or something, or possibly just be deleted (otherwise it will persist forever, likely even after userinfo normalization has been implemented :))